### DISCUSSION TITLE
Upload Workouts

### DISCUSSION THREAD

**Discussion by zarubin-alexey** [sealbro/dotnet.garmin.connect#64] [General]
Does Update Workout performs new workout upload? If not, we may want to add upload functionality.

**Comment by sealbro** [comment]
As for me adding new workouts is a good idea, but for the first step we should describe `WorkoutSteps` from `WorkoutSegments`. It depends on `SportType` because now we have a lot of `object` properties in `WorkoutSteps`. On the other hand, depends on other methods' properties like `id` or `description`
After documenting and replacing properties with strong types of `WorkoutSteps` we can add a new workout and improve the updated workout.

**Comment by zarubin-alexey** [comment]
Works as UpdateWorkout, requires WorkoutId = 0, headers should be empty (use Post).
It is enough to upload a new workout. Probably another method(not Update) should be used.
I was used to import workout from 1 user to another user.

**Reply by sealbro** [reply]
Great, If it works you can create PR 
In the final result the best test with all popular workouts like running, swimming, and cycling.

**Reply by macsimc** [reply]
This is what i'm currently using, just make sure you set Type="ExecutableStepDTO" or "RepeatGroupDTO" in GarminWorkoutStep

```
 public Task<GarminWorkout> CreateWorkout(GarminWorkout workout, CancellationToken cancellationToken = default)
    {
        return _context.PostAndDeserialize<GarminWorkout,GarminUpdateWorkout>(WorkoutUrl, GarminUpdateWorkout.From(workout), null, cancellationToken);
    }
```

and in GarminConnectContext.cs

 ```
  public async Task<T> PostAndDeserialize<T,TBody>(string url, TBody body,
       IReadOnlyDictionary<string, string> headers = null, CancellationToken cancellationToken = default)
   {
       var response = await MakeHttpPost(url,body,headers, cancellationToken: cancellationToken);
       if (response.StatusCode == HttpStatusCode.NoContent)
       {
           return default;
       }

       var json = await response.Content.ReadAsByteArrayAsync(cancellationToken);

       return GarminSerializer.To<T>(json);
   }
```

**Comment by Mrsevic** [comment]
Hi @sealbro and thank you for the library. 
I am wondering whether it'd be possible for you to increase the access modifiers for GarminUpdateWorkout and _context to protected internal. It'd make testing these use cases a bit easier.

Btw I will try to create a PR if all goes well.

Thanks
